### ISSUE TITLE
Post-training integer quantization for SSDMobileNet - poor detection accuracy

### ISSUE THREAD

**Issue by Garstig** [tensorflow/model-optimization#1002]
Hi there

**Describe the bug**
I used [Convert TF Object Detection API model to TFLite](https://colab.research.google.com/github/tensorflow/models/blob/master/research/object_detection/colab_tutorials/convert_odt_model_to_TFLite.ipynb) and tried to modify it, so it is fully int 8 quantized. [Here](https://colab.research.google.com/drive/1XuWniYM6q_0Uffp6SqSC5GUoS4Bd78RM?usp=sharing) is my version. It should completely run on its own in Colab. 

In the notebook I use two different functions to create the different representative datasets.  The first one uses 200 images from the coco dataset, the second one only 2. However, the resulting models seem to perform quite similarly poorly. Therefore I guess I did something wrong but can't find the mistake. 

You can see the results of the models at the very end of the colab but I also will add some Screenshots. 

**System information**

I didn't change the standard Installation on google Colab.

Python version:

**Describe the expected behavior**

The first generated tflite model should perform much better than the second and detect all easy cases.

**Describe the current behavior**
They both perform bad. 

**Code to reproduce the issue**
[Here](https://colab.research.google.com/drive/1XuWniYM6q_0Uffp6SqSC5GUoS4Bd78RM?usp=sharing) is my version. It should completely run on its own on Colab. 

**Screenshots**

**Model 2 outperforms Model 1**

Result Model 1

![image](https://user-images.githubusercontent.com/32436937/181012025-4f99a839-9615-4680-9ab8-5b554ed18bb6.png)

Result Model 2
![image](https://user-images.githubusercontent.com/32436937/181011661-55c42b85-6e94-4fee-800c-d52e3d8848d9.png)

**Both perform okay**

Result Model 1

![image](https://user-images.githubusercontent.com/32436937/181013006-c3354528-89cd-411b-bd2d-2435a49967b6.png)

Result Model 2

![image](https://user-images.githubusercontent.com/32436937/181012986-1c49681c-4524-482b-872e-8fd710a963fc.png)

**Both perform poorly**

Result Model 1 & 2
![image](https://user-images.githubusercontent.com/32436937/181012125-c887032a-1744-43c7-9768-75b521e15fa2.png)

**Additional context**
If you have any questions about my code don't hesitate to ask! Thank you very much.

**Comment by dansuh17** [comment]
Hi @Garstig, thanks for the detailed description.

One quick question about the model in the examples: is Model 2 the model quantized with 200 representative samples?

Selecting the representative dataset is a subtle art. Could you try the following and see if the performance improves?

* Use a subset of the training dataset (instead of the validation dataset) for the representative dataset.
* Use a little more than 200 samples (512, for instance).
* Shuffle the files before selecting the samples. I'm not sure how COCO dataset looks like, but if the files are ordered somehow and you're only selecting the first 200, it may not be representing the whole spectrum of possible inputs.

**Comment by Garstig** [comment]
Hi @dansuh17, thanks for your answer!

Model2 was the model with  only 2 representative samples. So yes, there were cases were the model with less representative samples performed better than the one with more. 

I just tried everything you suggested. Unfortunately the results are still quite bad. I even used the same images for testing as for the representative dataset, as I thought it should work best for them. 

The updated colab notebook can be [found here](https://colab.research.google.com/drive/1mZvmgc9v4HQPsBWPtqOTyTHqXaWUwoNx?usp=sharing).  The tflite models can found [here](https://drive.google.com/drive/folders/17K57mzX04bFjf4RTsTmrSqAj6pxs54GS?usp=sharing) together with the images I used (random sample of the coco train dataset. My notebook downloads them from google drive, so you don't need to download 18gb of files). 

Do you know if the function for creating my representative dataset is right? 


`def _representative_dataset_gen_full():
    root = 'images/repr/'
    pattern = "*.jpg"
    imagePaths = []
    for path, subdirs, files in os.walk(root):
        for name in files:
            if fnmatch(name, pattern):
                imagePaths.append(root + name)        
    for index,p in enumerate(imagePaths): 
            if index % 50 == 0:
              print(index)
            image = cv2.imread(p)
            image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB) 
            image = cv2.resize(image, (640, 640)) 
            image = image.astype("float") 
            image = np.expand_dims(image, axis=1) 
            image = image.reshape(1, 640, 640, 3)
            #minus mean and devide by std
            image = (image - 127.5)/127.5
            yield [image.astype("float32")]
`

**Comment by Garstig** [comment]
Hi @dansuh17 , do you have any chance too look into it again? Or do you know someone who could help? I would really appreciate that!

**Comment by rino20** [comment]
Your representative dataset generation function is not aligned with your preprocess_image_cv() function. 

especially, I think image = image * 255 is missing. what about just call the preprocess_image_cv() in _representative_dataset_gen_full ?

**Comment by Garstig** [comment]
Whoops, I closed it by accident. I just wanted to comment.
